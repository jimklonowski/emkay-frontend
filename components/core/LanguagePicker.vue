<template>
  <v-menu
    v-model="menu"
    :close-on-content-click="false"
    offset-y
  >
    <template #activator="{ on }">
      <v-btn :title="$t('language')" icon v-on="on">
        <v-icon v-text="icon" />
      </v-btn>
    </template>
    <v-card>
      <v-list>
        <v-list-item v-for="locale in availableLocales" :key="locale.code">
          <v-btn
            :disabled="$i18n.locale === locale.code"
            :outlined="$i18n.locale === locale.code"
            small
            text
            block
            nuxt
            @click="changeLocale(locale)"
          >
            {{ locale.name }}
          </v-btn>
        </v-list-item>
      </v-list>
    </v-card>
  </v-menu>
</template>

<script>
export default {
  data: () => ({
    icon: 'mdi-web',
    menu: false
  }),
  computed: {
    availableLocales: vm => vm.$i18n.locales
  },
  methods: {
    changeLocale (locale) {
      let langRoute = this.switchLocalePath(locale.code)
      // remove trailing slash if '/'
      langRoute = (langRoute === '/') ? langRoute : langRoute.replace(/\/$/, '')
      this.$router.replace(langRoute)
    }
  }
}
</script>
