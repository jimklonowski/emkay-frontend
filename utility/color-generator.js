import { interpolateViridis as defaultColorFunction } from 'd3-scale-chromatic'

function calculatePoint (i, intervalSize, colorRangeInfo) {
  const { colorStart, colorEnd, useEndAsStart } = colorRangeInfo
  return useEndAsStart ? (colorEnd - (i * intervalSize)) : (colorStart + (i * intervalSize))
}

/**
 * Generate and return an array of colors using D3's chromatic interpolation library
 */
export function interpolateColors (dataLength, colorScale = defaultColorFunction, colorRangeInfo = { colorStart: 0, colorEnd: 1, useEndAsStart: true }) {
  const { colorStart, colorEnd } = colorRangeInfo
  const colorRange = colorEnd - colorStart
  const intervalSize = colorRange / dataLength
  const colorArray = []

  for (let i = 0; i < dataLength; i++) {
    const colorPoint = calculatePoint(i, intervalSize, colorRangeInfo)
    colorArray.push(colorScale(colorPoint))
  }
  return colorArray
}
