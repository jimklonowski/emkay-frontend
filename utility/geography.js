import state_abbreviations from '@/assets/json/state_abbreviations'
import province_abbreviations from '@/assets/json/province_abbreviations'

// convert single object to array of { abbreviation, name }
export const states = Object.entries(state_abbreviations).map(([abbreviation, name]) => ({ abbreviation, name }))
export const provinces = Object.entries(province_abbreviations).map(([abbreviation, name]) => ({ abbreviation, name }))

// join both arrays so we can use array.find() to lookup abbreviation/name from a single source
const lookup = [...states, ...provinces]

/**
 * Given a state/province abbreviation, returns the state/province name
 * @param {*} abbreviation state or province abbreviation
 */
export function stateAbbreviationToName (abbreviation) {
  const result = lookup.find(x => x.abbreviation.toUpperCase() === abbreviation.toUpperCase())
  return result.name
}

/**
 * Given a state/province name, returns the state/province abbreviation
 * @param {*} name state or province name
 */
export function stateNameToAbbreviation (name) {
  const result = lookup.find(x => x.name.toUpperCase() === name.toUpperCase())
  return result.abbreviation
}

export function cityStateZip (city, state, zip) {
  const city_state = [city, state].filter(Boolean).join(', ')
  return [city_state, zip].filter(Boolean).join(' ')
}
