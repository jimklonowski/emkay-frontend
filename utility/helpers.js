/**
 * Click-to-Dial a phone number
 */
export const dialTo = (number) => { window.location = `tel:${number}` }

/**
 * Click-to-Compose an email
 */
export const emailTo = (email) => { window.location = `mailto:${email}` }

/**
 * Flatten tree-like center hierarchy
 */
export const flatten = array => array.reduce((res, { center_code, center_name, children = [] }) => res.concat({ center_code, center_name }).concat(flatten(children)), [])

/**
 * Format Currency
 */
export const formatCurrency = (value = 0, minDigits = 2, maxDigits = 2) => {
  const formatter = new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD', minimumFractionDigits: minDigits, maximumFractionDigits: maxDigits })
  return formatter.format(value)
}

/**
 * Use to add a delay in an async request (for debugging)
 * @param {*} m time in millis
 */
export const sleep = m => new Promise(resolve => setTimeout(resolve, m))

/**
 * Compute total by key for an array of objects
 * @param {*} items array of objects
 * @param {*} key key of object to sum
 */
export const computeTotalByKey = (items, key) => {
  const total = items.reduce((a, b) => {
    return {
      [key]: Number(a[key]) + Number(b[key])
    }
  }, { [key]: 0 })
  return total[key]
}

/**
 * Multifilter
 * https://gist.github.com/jherax/f11d669ba286f21b7a2dcff69621eb72
 */
export const multiFilter = (item, condition) => {
  const filterKeys = Object.keys(condition)
  return item.filter((eachObj) => {
    return filterKeys.every(eachKey => {
      if (!condition[eachKey].length) {
        return true // passing an empty filter means that filter is ignored
      }
      // this works, but is not case insensitive
      // return Array.prototype.includes.call(condition[eachKey], eachObj[eachKey].toString().toLowerCase())
      const lowercaseArray = condition[eachKey].map(x => x.toString().toLowerCase())
      return Array.prototype.includes.call(lowercaseArray, eachObj[eachKey].toString().toLowerCase())
    })
  })
}

/**
 * Object key comparator function
 */
export const compareObjectByKey = (key, order = 'asc') => {
  return function innerSort (a, b) {
    // eslint-disable-next-line no-prototype-builtins
    if (!a.hasOwnProperty(key) || !b.hasOwnProperty(key)) {
      // property doesn't exist on either object
      return 0
    }
    const varA = (typeof a[key] === 'string') ? a[key].toUpperCase() : a[key]
    const varB = (typeof b[key] === 'string') ? b[key].toUpperCase() : b[key]
    let comparison = 0
    if (varA > varB) { comparison = 1 } else if (varA < varB) { comparison = -1 }
    return ((order === 'desc') ? (comparison * -1) : comparison)
  }
}
