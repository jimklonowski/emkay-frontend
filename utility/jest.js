/**
 * Adds a warapping `div data-app="true"` to the body so that we don't
 * get Vuetify complaining about missing data-app attribute for some components.
 *
 * @return undefined
 */
export function addElemWithDataAppToBody () {
  const app = document.createElement('div')
  app.setAttribute('data-app', true)
  document.body.append(app)
}
