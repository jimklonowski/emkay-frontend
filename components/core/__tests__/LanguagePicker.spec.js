import { createLocalVue, mount } from '@vue/test-utils'
import Vuetify from 'vuetify'
import LanguagePicker from '../LanguagePicker'
import { addElemWithDataAppToBody } from '@/utility/jest'
const locales = require('@/plugins/i18n/locales')
const localVue = createLocalVue()
describe('LanguagePicker', () => {
  let vuetify
  let wrapper
  let menuButton
  let localeButtons

  beforeEach(async () => {
    vuetify = new Vuetify()
    wrapper = mount(LanguagePicker, {
      localVue,
      vuetify,
      mocks: {
        $i18n: { locale: 'en', locales }
      }
    })
    addElemWithDataAppToBody()
    menuButton = wrapper.find('.v-btn--icon')
    menuButton.trigger('click')
    await wrapper.vm.$nextTick()
    localeButtons = wrapper.findAll('.v-menu__content > .v-card > .v-list > .v-list-item > button')
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('mounts properly', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  it('renders properly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('opens menu on click', () => {
    expect(wrapper.vm.menu).toBe(true)
  })

  it('shows 3 locale buttons', () => {
    expect(localeButtons).toHaveLength(3)
  })

  it('calls changeLocale on locale button click', () => {
    const frenchButton = localeButtons.at(2)
    wrapper.vm.changeLocale = jest.fn()
    expect(wrapper.vm.changeLocale).toHaveBeenCalledTimes(0)
    frenchButton.trigger('click')
    expect(wrapper.vm.changeLocale).toHaveBeenCalledTimes(1)
  })
})
