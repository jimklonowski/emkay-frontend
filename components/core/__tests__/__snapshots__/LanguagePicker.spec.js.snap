// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguagePicker renders properly 1`] = `"<div class=\\"v-menu\\"><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round theme--light v-size--default\\" title=\\"language\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-web theme--light\\"></i></span></button></div>"`;
